base/stats.go

```diff
@@ -0,0 +1,799 @@
+package base
+
+import (
+	"expvar"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	StatViewFormat = "%s.%s"
+
+	NamespaceKey                 = "sgw"
+	ResourceUtilizationSubsystem = "resource_utilization"
+
+	SubsystemCacheKey           = "cache"
+	SubsystemDatabaseKey        = "database"
+	SubsystemDeltaSyncKey       = "delta_sync"
+	SubsystemGSIViews           = "gsi_views"
+	SubsystemReplication        = "replication"
+	SubsystemReplicationPull    = "replication_pull"
+	SubsystemReplicationPush    = "replication_push"
+	SubsystemSecurity           = "security"
+	SubsystemSharedBucketImport = "shared_bucket_import"
+)
+
+var (
+	checkedSentDesc               *prometheus.Desc
+	numAttachmentBytesTransferred *prometheus.Desc
+	numAttachmentsTransferred     *prometheus.Desc
+	numDocsFailedToPush           *prometheus.Desc
+	numDocsPushed                 *prometheus.Desc
+)
+
+type SgwStats struct {
+	GlobalStats     *GlobalStat         `json:"global"`
+	DbStats         map[string]*DbStats `json:"per_db"`
+	ReplicatorStats *ReplicatorStats    `json:"per_replication,omitempty"`
+
+	dbStatsMapMutex sync.Mutex
+}
+
+func NewSyncGatewayStats() *SgwStats {
+	sgwStats := SgwStats{
+		GlobalStats: &GlobalStat{},
+		DbStats:     map[string]*DbStats{},
+	}
+
+	checkedSentDesc = prometheus.NewDesc(prometheus.BuildFQName(NamespaceKey, SubsystemReplication, "sgr_docs_checked_sent"), "sgr_docs_checked_sent", []string{"replication"}, nil)
+	numAttachmentBytesTransferred = prometheus.NewDesc(prometheus.BuildFQName(NamespaceKey, SubsystemReplication, "sgr_num_attachment_bytes_transferred"), "sgr_num_attachment_bytes_transferred", []string{"replication"}, nil)
+	numAttachmentsTransferred = prometheus.NewDesc(prometheus.BuildFQName(NamespaceKey, SubsystemReplication, "sgr_num_attachments_transferred"), "sgr_num_attachments_transferred", []string{"replication"}, nil)
+	numDocsFailedToPush = prometheus.NewDesc(prometheus.BuildFQName(NamespaceKey, SubsystemReplication, "sgr_num_docs_failed_to_push"), "sgr_num_docs_failed_to_push", []string{"replication"}, nil)
+	numDocsPushed = prometheus.NewDesc(prometheus.BuildFQName(NamespaceKey, SubsystemReplication, "sgr_num_docs_pushed"), "sgr_num_docs_pushed", []string{"replication"}, nil)
+
+	sgwStats.GlobalStats.initResourceUtilizationStats()
+	sgwStats.initReplicationStats()
+
+	return &sgwStats
+}
+
+// This String() is to satisfy the expvar.Var interface which is used to produce the expvar endpoint output.
+func (s *SgwStats) String() string {
+	s.dbStatsMapMutex.Lock()
+	bytes, err := JSONMarshalCanonical(s)
```

Comment:
JsonMarshalCanonical seems like it's going to be quite a bit more expensive than the expvar.Map.String() we were using previously - in particular because expvar.Map sorts keys as they are added to the map, where this would be sorting the keys each time String() is called.  I'm also curious about memory allocation associated with json Marshal instead of expvar.Map.String().  If you haven't already, can you write a benchmark and generate some profiling to evaluate the performance difference?  Once we see the difference, we can decide whether it's a tolerable degradation for customers doing frequent  _expvar polling as part of their runtime diagnostics today.

Reply:
Done some benchmarks and have got some profiles ran and it would seem that for some reason the new implementation is actually better? We can chat about this in sync up.